# Descriptive Statistics Activity 2

Dataset: Data containing informations on car's name, origin, model, model year, cylinders etc.

Columns:
- mpg,
- cylinders,
- displacement,
- horsepower,
- weight,
- acceleration,
- model_year,
- origin,
- name

## Import libraries

We can calculate measures of central tendency and spread using `pandas dataframes`.

First thing you need to do is import `pandas` below.

### Step 1. &nbsp;&nbsp;&nbsp; Read CSV Data into Pandas Dataframe

#!blhint

- Create variable &nbsp;`pd`
- `import pandas as pd`

<br><details><summary>Blockly Hints</summary><ul><li>Blocks<ul><li><code>import .. as ..</code><ul><li>The <code>import .. as ..</code> block is found under IMPORT.</li> <li>A variable must have been created before it will appear in the dropdown.</li></ul></li></ul></li></ul></details>

In [1]:
#!response

import pandas as pd

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="zq]t,,UR`GWj6?Hub9%a">pd</variable></variables><block type="importAs_Python" id="o[3wY[w:R*b$p^Ow/ZZw" x="125" y="352"><field name="libraryName">pandas</field><field name="libraryAlias" id="zq]t,,UR`GWj6?Hub9%a">pd</field></block></xml>

#### Substep. &nbsp;&nbsp; Read CSV data and Save in Variable

We're going to work with data on cars. 

Load the CSV file `"datasets/mpg-na-hidden.csv"` into a dataframe and display it.

#!blhint

- Create variable &nbsp;`df`
- `set df to` &nbsp;:&nbsp; `with pd do read_csv using` &nbsp;<-&nbsp; `"datasets/mpg-na-hidden.csv"`

<details>
  <summary>Blockly Hints</summary>
  <ol>
    <li>Variable <code>dataframe</code> must be created under VARIABLES before it can be used</li>
    <li><code>set .. to</code> block found under VARIABLES</li>
    <li><code>with .. do .. using</code> block found under VARIABLES</li>
    <li>If <code>do</code> dropdown in <code>with .. do .. using</code> block will not populate, try "Run All Above Selected Cell" from the "Run" menu in top-left.</li>
    <li>You need a <code>".."</code> block found under TEXT to fill in the <code>using</code> part of the <code>with .. do .. using</code> block</li>
    <li>If <code>with .. do .. using</code> block does not want to snap together nicely with the <code>set .. to</code> block, try dragging the <code>set .. to</code> block instead.</li>
      <li>From <code>VARIABLES</code>, get the <code>dataframe</code> block</li>
  </ol>
</details>

In [2]:
#!response

dataframe = pd.read_csv('datasets/mpg-na-hidden.csv')

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="[V~uW+0L/4GW;45ulv+l">dataframe</variable><variable id="zq]t,,UR`GWj6?Hub9%a">pd</variable></variables><block type="variables_set" id="i*U..F)p9r]#n*e(./*x" x="0" y="0"><field name="VAR" id="[V~uW+0L/4GW;45ulv+l">dataframe</field><value name="VALUE"><block type="varDoMethod_Python" id=":~$/ssC/m-m@jj8K*O4W"><mutation items="1"></mutation><field name="VAR" id="zq]t,,UR`GWj6?Hub9%a">pd</field><field name="MEMBER">read_csv</field><data>pd:read_csv</data></block></value></block><block type="variables_set" id="Etf1jV+uJzk;:yf^gZ*9" x="19" y="228"><field name="VAR" id="[V~uW+0L/4GW;45ulv+l">dataframe</field><value name="VALUE"><block type="varDoMethod_Python" id="CAA$~5kuG7#(|]inmMix"><mutation items="1"></mutation><field name="VAR" id="zq]t,,UR`GWj6?Hub9%a">pd</field><field name="MEMBER">read_csv</field><data>pd:read_csv</data><value name="ADD0"><block type="text" id="@h?tZq_F}qks@eCoZDz7"><field name="TEXT">datasets/mpg-na-hidden.csv</field></block></value></block></value></block></xml>

In [3]:
dataframe

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="[V~uW+0L/4GW;45ulv+l">dataframe</variable></variables><block type="variables_get" id="10id|j5+R!4Pu9~lPyG`" x="8" y="346"><field name="VAR" id="[V~uW+0L/4GW;45ulv+l">dataframe</field></block></xml>

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0         130    3504          12.0   
1    15.0          8         350.0         165    3693          11.5   
2    18.0          8         318.0         150    3436          11.0   
3    16.0          8         304.0         150    3433          12.0   
4    17.0          8         302.0         140    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0          86    2790          15.6   
394  44.0          4          97.0          52    2130          24.6   
395  32.0          4         135.0          84    2295          11.6   
396  28.0          4         120.0          79    2625          18.6   
397  31.0          4         119.0          82    2720          19.4   

     model_year  origin                       name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[398 rows x 9 columns]

## Car mean/median/mode

In the next three cells, use the dataframe to calculate the mean, median, and mode values for the `mpg`, `cylinders`, `weight` and `acceleration` columns.

#### Substep. Compute the mean  


Blockly Hints:
- create variable `mean_comp`
- `Set mean_comp to` : `dataframe[ .. ]` <- `create list with`:
    - "mpg"
    - "cylinders"
    - "weight"
    - "acceleration"


In [4]:
#!response

mean_comp = dataframe[['mpg', 'cylinders', 'weight', 'acceleration']]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="5E_4,/m14pLKhp]%)#3Z">mean_comp</variable><variable id="[V~uW+0L/4GW;45ulv+l">dataframe</variable></variables><block type="variables_set" id=")7k9kW$|Z?7p/,{3kY+n" x="16" y="124"><field name="VAR" id="5E_4,/m14pLKhp]%)#3Z">mean_comp</field><value name="VALUE"><block type="indexer_Python" id="#)z2t%`3%`!UoW=w)do="><field name="VAR" id="[V~uW+0L/4GW;45ulv+l">dataframe</field><value name="INDEX"><block type="lists_create_with" id="M!HfZy3eg=Hk#j#R1gHp"><mutation items="4"></mutation><value name="ADD0"><block type="text" id="?vl9y/44e;a(.I-{?^a_"><field name="TEXT">mpg</field></block></value><value name="ADD1"><block type="text" id="-V17]]i;RX9d~8f~tX8I"><field name="TEXT">cylinders</field></block></value><value name="ADD2"><block type="text" id="w*={?^gK8+4HA$!TYvTA"><field name="TEXT">weight</field></block></value><value name="ADD3"><block type="text" id="P~Dd-.!4{rAL!wu2Y}go"><field name="TEXT">acceleration</field></block></value></block></value></block></value></block></xml>

Blockly Hints:
- `with mean_comp do mode using`

In [5]:
#!response

mean_comp.mean()

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="5E_4,/m14pLKhp]%)#3Z">mean_comp</variable></variables><block type="varDoMethod_Python" id="cldP}*$k{b1MeN[Km%}K" x="-402" y="-286"><mutation items="1"></mutation><field name="VAR" id="5E_4,/m14pLKhp]%)#3Z">mean_comp</field><field name="MEMBER">mean</field><data>mean_comp:mean</data></block></xml>

mpg               23.514573
cylinders          5.454774
weight          2970.424623
acceleration      15.568090
dtype: float64

### QUESTIONS:

### We have 9 observations of GRE scores.  314, 312, 300, 304, 323, 308, 316, 320, 298. What is the mean GRE score ?

#!response

Ans:

### In what scenario, the mean and mode of a variable will be same? Think deeply and give your answer. 

#!response


Ans: 

#### Substep. Compute the median  

Using the same approach, compute the median. 


Blockly Hints:
- create variable `median_comp`
- `Set median_comp to` : `dataframe[ .. ]` <- `create list with`:
    - "mpg"
    - "cylinders"
    - "weight"
    - "acceleration"


In [6]:
#!response

median_comp = dataframe[['mpg', 'cylinders', 'weight', 'acceleration']]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="p6RH-}?7)Y}V0wEwp5uK">median_comp</variable><variable id="[V~uW+0L/4GW;45ulv+l">dataframe</variable></variables><block type="variables_set" id="ylD)7lx|j`}lKJ-T)n5U" x="-210" y="-216"><field name="VAR" id="p6RH-}?7)Y}V0wEwp5uK">median_comp</field><value name="VALUE"><block type="indexer_Python" id=",xWp/CzM79Px#O5%6J+z"><field name="VAR" id="[V~uW+0L/4GW;45ulv+l">dataframe</field><value name="INDEX"><block type="lists_create_with" id="(OGvK0vP8[~0rPugVb#:"><mutation items="4"></mutation><value name="ADD0"><block type="text" id="dGq#ZsyrADc=U#UY4uX$"><field name="TEXT">mpg</field></block></value><value name="ADD1"><block type="text" id="l`DV322Z.tK8L5}^/_]]"><field name="TEXT">cylinders</field></block></value><value name="ADD2"><block type="text" id="vZBBE9b##0Gb4)y2?ER7"><field name="TEXT">weight</field></block></value><value name="ADD3"><block type="text" id="*_cPhLZ,,K8XMOt4jZf3"><field name="TEXT">acceleration</field></block></value></block></value></block></value></block></xml>

Blockly Hints:

- `with median_comp do mode using`

In [7]:
#!response

median_comp.median()

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="p6RH-}?7)Y}V0wEwp5uK">median_comp</variable></variables><block type="varDoMethod_Python" id="]I:]}wMu=(Gnu*RD$$~9" x="-124" y="-258"><mutation items="1"></mutation><field name="VAR" id="p6RH-}?7)Y}V0wEwp5uK">median_comp</field><field name="MEMBER">median</field><data>median_comp:median</data></block></xml>

mpg               23.0
cylinders          4.0
weight          2803.5
acceleration      15.5
dtype: float64

### QUESTIONS:

### What does the median value of cylinder means according to you? Explain briefly.

#!response

Ans:

### For a variable, what percentile value does its the median value refers to?

#!response

Ans: 

#### Substep. Compute the mode  

Using the same approach, compute the mode. 

Blockly Hints:
- create variable `mode_comp`
- `Set mode_comp to` : `dataframe[ .. ]` <- `create list with`:
    - "mpg"
    - "cylinders"
    - "weight"
    - "acceleration"


In [9]:
#!response

mode_comp = dataframe[['mpg', 'cylinders', 'weight', 'acceleration']]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="gOTFB1xZFtK`ZIe+EV`i">mode_comp</variable><variable id="[V~uW+0L/4GW;45ulv+l">dataframe</variable></variables><block type="variables_set" id="{1~gPZ^Roe]/4]O~aWvf" x="-188" y="-347"><field name="VAR" id="gOTFB1xZFtK`ZIe+EV`i">mode_comp</field><value name="VALUE"><block type="indexer_Python" id="c5@KJJWG9X6eU!KZstgk"><field name="VAR" id="[V~uW+0L/4GW;45ulv+l">dataframe</field><value name="INDEX"><block type="lists_create_with" id="?1c+#Ehh#Xlc-LIQ?FFu"><mutation items="4"></mutation><value name="ADD0"><block type="text" id="-lcp]t{)cMtkvki@f?iB"><field name="TEXT">mpg</field></block></value><value name="ADD1"><block type="text" id="E.hY[l)LCnfhrp1+J[Qc"><field name="TEXT">cylinders</field></block></value><value name="ADD2"><block type="text" id="^rk3dOI1LIkm?f+`$Q2-"><field name="TEXT">weight</field></block></value><value name="ADD3"><block type="text" id="E*cPX8.M]EPM^%%H7B9p"><field name="TEXT">acceleration</field></block></value></block></value></block></value></block></xml>

Blockly Hints:
- `with mode_comp do mode using`

In [11]:
#!response

mode_comp.mode()

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="gOTFB1xZFtK`ZIe+EV`i">mode_comp</variable></variables><block type="varDoMethod_Python" id="]I:]}wMu=(Gnu*RD$$~9" x="-175" y="-287"><mutation items="1"></mutation><field name="VAR" id="gOTFB1xZFtK`ZIe+EV`i">mode_comp</field><field name="MEMBER">mode</field><data>mode_comp:mode</data></block></xml>

mpg  cylinders  weight  acceleration
0  13.0        4.0    1985          14.5
1   NaN        NaN    2130           NaN

### QUESTIONS:

### Suppose we have a variable 'Age'. It has 50 values. We find its  mode value is 26. What does that mean? Briefly Explain.

#!response


Ans:

### What is the significance of the two rows `0` and `1` in the mode table?  

Write your answer in the cell below.

#!response

*Everything is numeric (actually ordinal), so it is the only variable Python will calculate mean and median for. 
It actually is incorrect to calculate mean because it is ordinal, but Python isn't smart enough to know that because we didn't tell it our data was ordinal. 
This is a good example of being careful with computers - they will do exactly what we tell them to do even if it is incorrect. 
It is up to us to make sure we understand the reasoning behind what we are doing so we can recognize when the computer gives us garbage answers.*